### PR TITLE
feat: Support Binary in shuffle writer

### LINKED ISSUE #105: data type Binary not supported in shuffle write
### Describe the bug

Found the issue when `CometCollectExec` is added during #100.

The error msg is: `data type Binary not supported in shuffle write`

Reproduced in the local environment, the whole stack trace is :
```
org.apache.comet.CometNativeException: not implemented: data type Binary not supported in shuffle write
        at std::backtrace_rs::backtrace::libunwind::trace(/rustc/ec08a0337f3556212525dbf1d3b41e19bdf27621/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93)
        at std::backtrace_rs::backtrace::trace_unsynchronized(/rustc/ec08a0337f3556212525dbf1d3b41e19bdf27621/library/std/src/../../backtrace/src/backtrace/mod.rs:66)
        at std::backtrace::Backtrace::create(/rustc/ec08a0337f3556212525dbf1d3b41e19bdf27621/library/std/src/backtrace.rs:331)
        at comet::errors::init::{{closure}}(./core/src/errors.rs:134)
        at  as core::ops::function::Fn>::call(/rustc/ec08a0337f3556212525dbf1d3b41e19bdf27621/library/alloc/src/boxed.rs:2021)
        at std::panicking::rust_panic_with_hook(/rustc/ec08a0337f3556212525dbf1d3b41e19bdf27621/library/std/src/panicking.rs:757)
        at std::panicking::begin_panic_handler::{{closure}}(/rustc/ec08a0337f3556212525dbf1d3b41e19bdf27621/library/std/src/panicking.rs:631)
        at std::sys_common::backtrace::__rust_end_short_backtrace(/rustc/ec08a0337f3556212525dbf1d3b41e19bdf27621/library/std/src/sys_common/backtrace.rs:170)
        at rust_begin_unwind(/rustc/ec08a0337f3556212525dbf1d3b41e19bdf27621/library/std/src/panicking.rs:619)
        at core::panicking::panic_fmt(/rustc/ec08a0337f3556212525dbf1d3b41e19bdf27621/library/core/src/panicking.rs:72)
        at comet::execution::datafusion::shuffle_writer::slot_size(./core/src/execution/datafusion/shuffle_writer.rs:319)
        at comet::execution::datafusion::shuffle_writer::PartitionBuffer::init_active_if_necessary::{{closure}}(./core/src/execution/datafusion/shuffle_writer.rs:225)
        at core::iter::adapters::map::map_fold::{{closure}}(/rustc/ec08a0337f3556212525dbf1d3b41e19bdf27621/library/core/src/iter/adapters/map.rs:84)
        at core::iter::traits::iterator::Iterator::fold(/rustc/ec08a0337f3556212525dbf1d3b41e19bdf27621/library/core/src/iter/traits/iterator.rs:2639)
        at  as core::iter::traits::iterator::Iterator>::fold(/rustc/ec08a0337f3556212525dbf1d3b41e19bdf27621/library/core/src/iter/adapters/map.rs:124)
        at ::sum(/rustc/ec08a0337f3556212525dbf1d3b41e19bdf27621/library/core/src/iter/traits/accum.rs:50)
        at core::iter::traits::iterator::Iterator::sum(/rustc/ec08a0337f3556212525dbf1d3b41e19bdf27621/library/core/src/iter/traits/iterator.rs:3634)
        at comet::execution::datafusion::shuffle_writer::PartitionBuffer::init_active_if_necessary(./core/src/execution/datafusion/shuffle_writer.rs:221)
        at comet::execution::datafusion::shuffle_writer::PartitionBuffer::append_rows(./core/src/execution/datafusion/shuffle_writer.rs:246)
        at comet::execution::datafusion::shuffle_writer::PartitionBuffer::append_batch(./core/src/execution/datafusion/shuffle_writer.rs:237)
        at comet::execution::datafusion::shuffle_writer::ShuffleRepartitioner::insert_batch::{{closure}}(./core/src/execution/datafusion/shuffle_writer.rs:730)
        at comet::execution::datafusion::shuffle_writer::external_shuffle::{{closure}}(./core/src/execution/datafusion/shuffle_writer.rs:930)
```

It's clear that `slot_size` and `append_columns` doesn't include binary support.

### Steps to reproduce

_No response_

### Expected behavior

Binary data could be wrote by shuffle writer.

### Additional context

I can submit a PR for this.

### PR DESCRIPTION
**Pull request by advancedxy**
## Which issue does this PR close?
Closes #105 

## Rationale for this change
bug fixes

## What changes are included in this PR?
Add binary pattern matching in shuffle_writer.rs

## How are these changes tested?
Add new test cases in rust and spark side

### CODE REVIEW

File: core/src/execution/datafusion/shuffle_writer.rs [hunk 1]
```diff
@@ -315,6 +315,9 @@ fn slot_size(len: usize, data_type: &DataType) -> usize {
             // TODO: this is not accurate, but should be good enough for now
             slot_size(len, key_type.as_ref()) + slot_size(len / 10, value_type.as_ref())
         }
+        // TODO: this is not accurate, but should be good enough for now
+        DataType::Binary => len * 100 + len * 4,
+        DataType::LargeBinary => len * 100 + len * 8,
```
Comment (viirya) [inline core/src/execution/datafusion/shuffle_writer.rs:320]: Oh, I missed binary type.

File: core/src/execution/datafusion/shuffle_writer/shuffle_writer_test.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! unit tests for the shuffle writer
+
+use super::*;
+
+#[test]
+fn test_slot_size() {
```
Comment (viirya) [inline core/src/execution/datafusion/shuffle_writer/shuffle_writer_test.rs:23]: Could you put this unit test in `shuffle_writer.rs` as usual?

**Comment by viirya** [comment]
> I'm not sure why this issue doesn't happen before... Seems like binary data are already used in various shuffle suites.

They use columnar shuffle, I think, which has much more coverage.

File: core/src/execution/datafusion/shuffle_writer/shuffle_writer_test.rs (hunk 2 above)
Comment (viirya) [reply core/src/execution/datafusion/shuffle_writer/shuffle_writer_test.rs:23]: Usually the unit test code is in same file in Rust as I see. You can see we also do it for other Rust files.

**Comment by viirya** [comment]
The TPCDS CI pipeline failure is going to be fixed at https://github.com/apache/arrow-datafusion-comet/pull/108.

**Review by sunchao — APPROVED** [review]
LGTM

**Comment by sunchao** [comment]
Merged, thanks!

**Pull request merged**